### PR TITLE
zasm: 4.2.6 -> 4.4.7, fix darwin build and cross-compilation

### PR DESCRIPTION
**Pull request by siraben**
###### Motivation for this change
Package version bump and cross-compilation fix (test with `nix build .#pkgsCross.raspberryPi.zasm`)
@SuperSandro2000 please check to see if it's still broken on aarch64.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 113402` at cf708951 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/87c693f0409ca2426ac6e2a76d4d94b5">zasm</a></li></li>
  </ul>
</details>
<details>
  <summary>2 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`configurePhase` should probably contain `runHook preConfigure` and `runHook postConfigure`.

Near pkgs/development/compilers/zasm/default.nix:24:3:
```
   |
24 |   configurePhase = ''
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/development/compilers/zasm/default.nix:35:3:
```
   |
35 |   installPhase = ''
   |   ^
```
</li>
  </ul>
</details>

**Comment by siraben** [comment]
It builds for me on darwin.

File: pkgs/development/compilers/zasm/default.nix [hunk 1]
```diff
@@ -1,20 +1,22 @@
...
-  version = "4.2.6";
+  version = "4.4.7";
+
   src = fetchFromGitHub {
     owner = "megatokio";
     repo = "zasm";
-    rev = "f1424add17a5514895a598d6b5e3982579961519";
-    sha256 = "1kqnqdqp2bfsazs6vfx2aiqanxxagn8plx8g6rc11vmr8yqnnpks";
+    rev = version;
+    sha256 = "sha256-XBf852JYplyw3QDyPKTFw+THaL8c7Cm5pFhTVbeh2FE=";
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/zasm/default.nix:19]:
```suggestion
    sha256 = "0rpz9x62njvm7fr4x2qc6w9didb5glx5f3amnb3bwjrjj23dxxir";
```

**Comment by turboMaCk** [comment]
@GrahamcOfBorg build zasm

File: pkgs/development/compilers/zasm/default.nix [hunk 2]
```diff
@@ -23,22 +25,23 @@ stdenv.mkDerivation {
...
 
   meta = with lib; {
-    description = "Z80 / 8080 assembler (for unix-style OS)";
+    description = "Z80 / 8080 / Z180 assembler (for unix-style OS)";
     homepage = "https://k1.spdns.de/Develop/Projects/zasm/Distributions/";
     license = licenses.bsd2;
     maintainers = [ maintainers.turbomack ];
-    platforms = platforms.linux;
+    platforms = platforms.unix;
     badPlatforms = platforms.aarch64;
```
Comment (turboMaCk) [inline pkgs/development/compilers/zasm/default.nix:45]: if you want to try aarch please remove the badPlatform from here and we can try it using OfBorg build

**Comment by turboMaCk** [comment]
[x] Build for me on NixOS x86-64

**Comment by turboMaCk** [comment]
@siraben Does this build for you still on MacOS? I tried to build this via nixpkgs-review on Darwin and it failed for me.

```
❯❯❯  nixpkgs-review pr 113402 --eval local
$ git -c fetch.prune=false fetch --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0 pull/113402/head:refs/nixpkgs-review/1
remote: Enumerating objects: 7, done.
remote: Counting objects: 100% (7/7), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 7 (delta 4), reused 7 (delta 4), pack-reused 0
Unpacking objects: 100% (7/7), 1.10 KiB | 188.00 KiB/s, done.
From https://github.com/NixOS/nixpkgs
   478c52b7e54..f1bb6330342  master                -> refs/nixpkgs-review/0
 + c4a3e021a9a...a4e2d7e0c60 refs/pull/113402/head -> refs/nixpkgs-review/1  (forced update)
$ git worktree add /Users/marek/.cache/nixpkgs-review/pr-113402/nixpkgs f1bb633034222cc151d92dffe9dba438016c9e68
Preparing worktree (detached HEAD f1bb6330342)
Updating files: 100% (24454/24454), done.
HEAD is now at f1bb6330342 Merge pull request #113378 from siraben/knightos-build-graph-reduction
$ nix-env -f /Users/marek/.cache/nixpkgs-review/pr-113402/nixpkgs -qaP --xml --out-path --show-trace
$ git merge --no-commit a4e2d7e0c60ae6c27affd9073d1dac3b13497fe0
Automatic merge went well; stopped before committing as requested
$ nix-env -f /Users/marek/.cache/nixpkgs-review/pr-113402/nixpkgs -qaP --xml --out-path --show-trace --meta
1 package added:
zasm (init at 4.4.7)

warning: ignoring the user-specified setting 'experimental-features', because it is a restricted setting and you are not a trusted user
$ nix --experimental-features nix-command build --no-link --keep-going --option build-use-sandbox relaxed -f /Users/marek/.cache/nixpkgs-review/pr-113402/build.nix
warning: ignoring the user-specified setting 'experimental-features', because it is a restricted setting and you are not a trusted user
warning: ignoring the user-specified setting 'sandbox', because it is a restricted setting and you are not a trusted user
hash mismatch in fixed-output derivation '/nix/store/0dn4khwiycs6izfdxxfywjq1lk18904x-source':
  wanted: sha256:0rpz9x62njvm7fr4x2qc6w9didb5glx5f3amnb3bwjrjj23dxxir
  got:    sha256:0lfql6vmalsqljwjkv0wpxlcgr63qnj3rwh0vnq5r9jqcbkzq5sw
cannot build derivation '/nix/store/yix1xyra1izkiz5jp4j78k87wrkamm9w-zasm-4.4.7.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/258saj44qw17qdwaz8cbbdh21jf3dvr2-env.drv': 1 dependencies couldn't be built
error: --- Error ------------------------------------------------------------------------------------------------------------------------------------------------------- nix
build of '/nix/store/258saj44qw17qdwaz8cbbdh21jf3dvr2-env.drv' failed

Link to currently reviewing PR:
https://github.com/NixOS/nixpkgs/pull/113402

1 package failed to build:
zasm

warning: ignoring the user-specified setting 'experimental-features', because it is a restricted setting and you are not a trusted user
error: --- Error ------------------------------------------------------------------------------------------------------------------------------------------------------- nix
build log of '/nix/store/yix1xyra1izkiz5jp4j78k87wrkamm9w-zasm-4.4.7.drv' is not available
$ nix-shell /Users/marek/.cache/nixpkgs-review/pr-113402/shell.nix
```

I think the change of sha256 fixed NixOS but broke MacOS?

**Comment by siraben** [comment]
> @siraben Does this build for you still on MacOS? I tried to build this via nixpkgs-review on Darwin and it failed for me.
> 
> I think the change of sha256 fixed NixOS but broke MacOS?

You are right, the build fails for me now

```
➜  nixpkgs git:(zasm-xcompile) ✗ nix build .#zasm; tb
error: hash mismatch in fixed-output derivation '/nix/store/pnrhmr65hbrckdybzfay81cc2rf7whch-source.drv':
         specified: sha256-OfbehpAyS77GslUNVzp9ZbXYEjcMi06yO3VLK0xP/2Y=
            got:    sha256-XBf852JYplyw3QDyPKTFw+THaL8c7Cm5pFhTVbeh2FE=
error: 1 dependencies of derivation '/nix/store/yix1xyra1izkiz5jp4j78k87wrkamm9w-zasm-4.4.7.drv' failed to build
```

**Comment by turboMaCk** [comment]
I vaguely remember that I had some issues with this as well when I was originally packaging this and that's why I ended up restricting platforms so much. Anyway I think we should try to make this work at least for x86 Darwin and Linux.

**Comment by siraben** [comment]
Perhaps fetching the tarball will be more deterministic?

**Comment by siraben** [comment]
I believe I have a compromise. `libs-src` will be fetched with `fetchFromGitHub` but `src` will be fetched with `fetchurl`.

**Comment by siraben** [comment]
Correction: that will not solve the issue since 1. GitHub changes the tarball generation from time to time and 2. fetchFromGitHub already does a fetchzip.

File: pkgs/development/compilers/zasm/default.nix [hunk 3]
```diff
@@ -1,20 +1,22 @@
...
-  src = fetchFromGitHub {
-    owner = "megatokio";
-    repo = "zasm";
-    rev = "f1424add17a5514895a598d6b5e3982579961519";
-    sha256 = "1kqnqdqp2bfsazs6vfx2aiqanxxagn8plx8g6rc11vmr8yqnnpks";
+  version = "4.4.7";
+
+  # do not use fetchFromGitHub since that results in different sha256
+  # hashes on macOS and Linux
+  src = fetchurl {
```
Comment (turboMaCk) [inline pkgs/development/compilers/zasm/default.nix:17]: I think we should try either `builtins.fetchTarball` or `builtins.fetchGit`

**Comment by siraben** [comment]
Ok should be good now. The culprit was the `Test` folder in the repository which had files with weird names.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 113402` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>zasm</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

zasm:
<details>


`configurePhase` should probably contain `runHook preConfigure` and `runHook postConfigure`.

Near pkgs/development/compilers/zasm/default.nix:24:3:

```
   |
24 |   configurePhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/development/compilers/zasm/default.nix:35:3:

```
   |
35 |   installPhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
</details>

**Comment by SuperSandro2000** [comment]
@ofborg eval

**Comment by SuperSandro2000** [comment]
NIXPKGS_ALLOW_UNSUPPORTED_SYSTEM=1 nix-build . -A pkgsCross.aarch64-multiplatform.zasm

```
configuring
building
g++ -c -m64      -pipe -Os -Wall -DNDEBUG -DQT_NO_DEBUG -D_REENTRANT -std=c++11 -Wno-multichar -I. -I.. -I../Source -I../Libraries -o Error.o ../Source/Error.cpp
make: g++: No such file or directory
make: *** [Makefile:64: Error.o] Error 127
builder for '/nix/store/vds63bafys60fdrb4i37nijmx1sw5q9x-zasm-4.2.6-aarch64-unknown-linux-gnu.drv' failed with exit code 2
error: build of '/nix/store/vds63bafys60fdrb4i37nijmx1sw5q9x-zasm-4.2.6-aarch64-unknown-linux-gnu.drv' failed
```

**Review by turboMaCk — APPROVED** [review]
good work LGTM.

**Pull request merged**